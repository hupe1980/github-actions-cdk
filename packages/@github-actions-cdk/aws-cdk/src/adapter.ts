import { Annotations, Aspects } from "aws-cdk-lib";
import type { Construct, IConstruct } from "constructs";
import {
  AnnotationMetadataEntryType,
  Project,
  type ProjectProps,
  isValidationError,
} from "github-actions-cdk";

/**
 * The `AwsCdkAdapter` class integrates GitHub Actions workflows with AWS CDK constructs,
 * inheriting from the `Project` base class in `github-actions-cdk`.
 *
 * This adapter binds the lifecycle of a GitHub Actions workflow to an AWS CDK Construct,
 * allowing workflow creation, error handling, and annotation of errors and warnings
 * during the CDK synthesis process.
 */
export class AwsCdkAdapter extends Project {
  private readonly awsCdkScope: IConstruct;
  private hasValidationErrors: boolean;

  /**
   * Constructs a new `AwsCdkAdapter` instance.
   *
   * @param awsCdkScope - The AWS CDK construct scope associated with this adapter.
   * This scope is used as a base for adding validations, annotations, and managing synthesis errors.
   * @param props - Project properties for configuring GitHub Actions workflows.
   */
  constructor(awsCdkScope: Construct, props: ProjectProps = {}) {
    super(props);

    this.awsCdkScope = awsCdkScope;
    this.hasValidationErrors = false;

    // Add an aspect to automatically synthesize workflows within the CDK scope.
    Aspects.of(this.awsCdkScope).add({
      visit: (node: IConstruct) => {
        if (node === this.awsCdkScope) {
          this.synth();
        }
      },
    });
  }

  /**
   * Handles synthesis errors encountered during workflow generation.
   * If the error is a validation error, it registers the error message as a validation
   * message on the associated CDK scope.
   *
   * @param error - The error encountered during synthesis.
   * @throws Error - If the error is not a validation error, it will be re-thrown.
   */
  protected handleSynthesisError(error: unknown): void {
    if (isValidationError(error)) {
      this.hasValidationErrors = true;
      this.awsCdkScope.node.addValidation({
        validate: () => {
          return error.errors.map(({ message, source }) => `- [${source.node.path}]: ${message}`);
        },
      });
    } else {
      throw error;
    }
  }

  /**
   * Finalizes the synthesis process by transferring workflow annotations to
   * the CDK context as appropriate.
   *
   * This method checks each annotation's severity level (info, warning, error) and
   * adds it to the CDK scope using the `Annotations` utility.
   *
   * Additionally, this method stops synthesis if there are blocking errors,
   * unless overridden by `continueOnErrorAnnotations`.
   */
  protected finalizeSynthesis(): void {
    const workflows = Object.values(this.manifest.workflows);

    for (const workflow of workflows) {
      for (const annotation of workflow.annotations) {
        switch (annotation.level) {
          case AnnotationMetadataEntryType.INFO:
            Annotations.of(this.awsCdkScope).addInfo(annotation.message);
            break;
          case AnnotationMetadataEntryType.WARN:
            Annotations.of(this.awsCdkScope).addWarning(annotation.message);
            break;
          case AnnotationMetadataEntryType.ERROR:
            Annotations.of(this.awsCdkScope).addError(annotation.message);
            break;
          default:
            throw new Error(`Unknown annotation level: ${annotation.level}`);
        }
      }
    }

    // Halt synthesis if errors exist and should not be ignored.
    if (!this.continueOnErrorAnnotations && this.manifest.hasErrorAnnotation()) {
      return;
    }

    // Halt synthesis if validation errors are present.
    if (this.hasValidationErrors) {
      return;
    }

    // Add informational message upon successful synthesis of workflows.
    Annotations.of(this.awsCdkScope).addInfo(
      `GitHub Actions workflows generated at ${this.outdir}`,
    );
  }
}
