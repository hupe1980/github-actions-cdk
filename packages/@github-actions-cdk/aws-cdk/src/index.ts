export * from "./adapter";
export * from "./aws-credentials";
export * from "./oidc";
export * from "./pipeline";
export * from "./publish";
export * from "./synth";
