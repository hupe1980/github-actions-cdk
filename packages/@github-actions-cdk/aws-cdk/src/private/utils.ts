import * as path from "node:path";

/**
 * Converts a given path to POSIX format (using forward slashes).
 *
 * @param p - The input path.
 * @returns The path in POSIX format.
 */
export function posixPath(p: string) {
  return p.split(path.sep).join(path.posix.sep);
}
