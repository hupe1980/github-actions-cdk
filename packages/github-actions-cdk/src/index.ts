export * from "./action";
export * from "./annotations";
export * from "./aspect";
export * from "./check";
export * from "./common";
export * from "./component";
export * from "./concurrency";
export * from "./cron";
export * from "./defaults";
export * from "./expression";
export * from "./job";
export * from "./manifest";
export * from "./permissions";
export * from "./project";
export * from "./rules";
export * from "./step";
export * from "./synthesizer";
export * from "./trigger";
export * from "./validator";
export * from "./workflow";

// export submobules
export * as actions from "./actions";
export * as checks from "./checks";
