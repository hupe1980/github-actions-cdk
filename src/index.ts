export * from "./action";
export * from "./concurrency";
export * from "./defaults";
export * from "./job";
export * from "./permissions";
export * from "./project";
export * from "./step";
export * from "./workflow";
