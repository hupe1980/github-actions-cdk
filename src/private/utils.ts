import * as fs from "node:fs";

export function decamelize(
	str: string,
	options: { separator?: string } = {},
): string {
	const separator = options.separator || "_";
	return str
		.replace(/([a-z\d])([A-Z])/g, `$1${separator}$2`)
		.replace(/([A-Z]+)([A-Z][a-z\d]+)/g, `$1${separator}$2`)
		.toLowerCase();
}

export function snakeCaseKeys<T>(obj: T, sep = "-"): T {
	if (typeof obj !== "object" || obj == null) {
		return obj;
	}

	if (Array.isArray(obj)) {
		return obj.map((o) => snakeCaseKeys(o, sep)) as T;
	}

	const result: Record<string, unknown> = {};
	for (let [k, v] of Object.entries(obj)) {
		// we don't want to snake case environment variables
		if (k !== "env" && typeof v === "object" && v != null) {
			v = snakeCaseKeys(v, sep);
		}
		result[decamelize(k, { separator: sep })] = v;
	}
	return result as T;
}

export function ensureDirSync(outdir: string): void {
	if (!fs.existsSync(outdir)) {
		fs.mkdirSync(outdir, { recursive: true });
	}
}

export function cleanObject<T extends Record<string, unknown>>(
	obj?: T,
): T | undefined {
	if (obj === undefined) {
		return undefined;
	}

	const cleanedObj: Partial<T> = {};

	for (const [key, value] of Object.entries(obj)) {
		if (value !== undefined) {
			(cleanedObj as Record<string, unknown>)[key] = value;
		}
	}

	return Object.keys(cleanedObj).length > 0 ? (cleanedObj as T) : undefined;
}
