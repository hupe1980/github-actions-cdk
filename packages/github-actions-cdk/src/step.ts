import type { IConstruct } from "constructs";
import type { ShellType } from "./common";
import { Component } from "./component";
import { cleanObject } from "./private/utils";
import { RegularStepValidator, RunStepValidator, StepBaseValidator } from "./validator";

// Unique symbols to mark instances of StepBase, RunStep, and RegularStep
const STEP_BASE_SYMBOL = Symbol.for("github-actions-cdk.StepBase");
const RUN_STEP_SYMBOL = Symbol.for("github-actions-cdk.RunStep");
const REGULAR_STEP_SYMBOL = Symbol.for("github-actions-cdk.RegularStep");

/**
 * Common properties for all step types in a GitHub Actions workflow job.
 */
export interface CommonStepProps {
  /**
   * A descriptive name for the step, displayed in the GitHub Actions UI.
   * @example "Install dependencies"
   */
  readonly name?: string;

  /**
   * Conditional expression to determine if the step should run (equivalent to `if` in GitHub Actions).
   * Supports GitHub Actions expressions, e.g., `${{ success() }}`.
   * @example "${{ github.event_name == 'push' }}"
   */
  readonly condition?: string;

  /**
   * Maximum execution time for the step, in minutes.
   * @example 10
   */
  readonly timeoutMinutes?: number;

  /**
   * Whether the job should continue if this step fails.
   * @default false
   */
  readonly continueOnError?: boolean;

  /**
   * Environment variables specific to this step, overriding job-level or workflow-level variables.
   * @example { "NODE_ENV": "production" }
   */
  readonly env?: Record<string, string>;
}

/**
 * Configuration for a step that runs a shell command.
 */
export interface RunStepProps extends CommonStepProps {
  /**
   * Commands or scripts to execute in this step.
   * @example ["npm", "install"]
   */
  readonly run: string | string[];

  /**
   * Directory in which the step's command or action executes.
   * Defaults to the job's working directory if not specified.
   * @example "src/"
   */
  readonly workingDirectory?: string;

  /**
   * Shell environment for this step, allowing custom shells like `bash`, `pwsh`, etc.
   * @default "bash"
   * @example "pwsh"
   */
  readonly shell?: ShellType;
}

/**
 * Configuration for a step that uses a predefined GitHub Action.
 */
export interface RegularStepProps extends CommonStepProps {
  /**
   * GitHub Action to run, identified by repository or Docker image reference.
   * @example "actions/checkout@v2"
   */
  readonly uses: string;

  /**
   * Input parameters for the action, passed as a key-value map.
   * @example { "token": "${{ secrets.GITHUB_TOKEN }}" }
   */
  readonly parameters?: Record<string, unknown>;
}

/**
 * Base class representing a single step within a GitHub Actions job.
 */
export abstract class StepBase extends Component {
  /**
   * Checks if an object is an instance of `StepBase`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `StepBase`; otherwise, `false`.
   */
  public static isStepBase(x: unknown): x is StepBase {
    return x !== null && typeof x === "object" && STEP_BASE_SYMBOL in x;
  }

  public readonly name?: string; // Name of the step
  public readonly condition?: string; // Conditional execution expression
  public readonly env?: Record<string, string>; // Environment variables for this step
  public readonly continueOnError?: boolean; // Continue on error flag
  public readonly timeoutMinutes?: number; // Step timeout in minutes

  /**
   * Creates a new `StepBase` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this step.
   */
  constructor(scope: IConstruct, id: string, props: CommonStepProps) {
    super(scope, id);

    // Mark the construct scope with STEP_BASE_SYMBOL to denote it's a StepBase
    Object.defineProperty(this, STEP_BASE_SYMBOL, { value: true });

    this.name = props.name;
    this.condition = props.condition;
    this.env = props.env;
    this.continueOnError = props.continueOnError;
    this.timeoutMinutes = props.timeoutMinutes;

    this.node.addValidation(new StepBaseValidator(this));
  }

  /**
   * Retrieves the step's unique identifier within the context of a workflow job.
   * @returns The unique identifier for this step.
   */
  public get id(): string {
    return this.node.id;
  }

  /**
   * Generates the GitHub Actions output expression for this step.
   *
   * This method constructs a formatted string that represents the GitHub Actions
   * output expression for the specified output name. The returned string can be
   * used in workflows to reference outputs from this step in a reusable format.
   *
   * @param name - The name of the specific output to reference from this step.
   * @returns The full GitHub Actions expression for accessing this output, e.g.,
   *          "${{ steps.stepId.outputs.outputName }}".
   */
  public outputExpression(name: string): string {
    return `\${{ steps.${this.id}.outputs.${name} }}`;
  }
}

/**
 * Step that runs shell commands in a GitHub Actions job.
 */
export class RunStep extends StepBase {
  /**
   * Checks if an object is an instance of `RunStep`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `RunStep`; otherwise, `false`.
   */
  public static isRunStep(x: unknown): x is RunStep {
    return x !== null && typeof x === "object" && RUN_STEP_SYMBOL in x;
  }

  public readonly run: string[]; // Commands to execute in the step
  public readonly workingDirectory?: string; // Directory for execution
  public readonly shell?: ShellType; // Shell type for execution

  /**
   * Creates a new `RunStep` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this run step.
   */
  constructor(scope: IConstruct, id: string, props: RunStepProps) {
    super(scope, id, props);

    Object.defineProperty(this, RUN_STEP_SYMBOL, { value: true });

    // Ensure the run property is always an array
    this.run = typeof props.run === "string" ? [props.run] : props.run;
    this.workingDirectory = props.workingDirectory;
    this.shell = props.shell;

    // Validation for the shell type
    this.node.addValidation(new RunStepValidator(this));
  }

  /**
   * Serializes the `RunStep` configuration to an object for GitHub Actions YAML output.
   *
   * Converts the run step configuration into an object format suitable for
   * GitHub Actions YAML serialization.
   *
   * @returns An object suitable for YAML serialization.
   * @internal
   * @override
   */
  public _toRecord(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      if: this.condition,
      run: this.run.join("\n"),
      env: this.env,
      continueOnError: this.continueOnError,
      timeoutMinutes: this.timeoutMinutes,
      workingDirectory: this.workingDirectory,
      shell: this.shell,
    };
  }
}

/**
 * Step that runs a predefined GitHub Action within a job.
 */
export class RegularStep extends StepBase {
  /**
   * Checks if an object is an instance of `RegularStep`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `RegularStep`; otherwise, `false`.
   */
  public static isRegularStep(x: unknown): x is RegularStep {
    return x !== null && typeof x === "object" && REGULAR_STEP_SYMBOL in x;
  }

  public readonly uses: string; // Reference to the GitHub Action
  public readonly parameters?: Record<string, unknown>; // Input parameters for the action

  /**
   * Creates a new `RegularStep` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this regular step.
   */
  constructor(scope: IConstruct, id: string, props: RegularStepProps) {
    super(scope, id, props);

    Object.defineProperty(this, REGULAR_STEP_SYMBOL, { value: true });

    this.uses = props.uses;
    this.parameters = props.parameters;

    this.node.addValidation(new RegularStepValidator(this));
  }

  /**
   * Determines if the action is an external action.
   *
   * @returns `true` if the action is an external action; otherwise, `false`.
   */
  public isExternalAction(): boolean {
    return !this.isRepoAction() && !this.isDockerAction();
  }

  /**
   * Determines if the action is a repository action.
   *
   * @returns `true` if the action is a repository action; otherwise, `false`.
   */
  public isRepoAction(): boolean {
    return this.uses.startsWith("./");
  }

  /**
   * Determines if the action is a Docker action.
   *
   * @returns `true` if the action is a Docker action; otherwise, `false`.
   */
  public isDockerAction(): boolean {
    return this.uses.startsWith("docker://");
  }

  /**
   * Serializes the `RegularStep` configuration to an object for GitHub Actions YAML output.
   *
   * Converts the step configuration into an object format suitable for
   * GitHub Actions YAML serialization.
   *
   * @returns An object suitable for YAML serialization.
   * @internal
   * @override
   */
  public _toRecord(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      if: this.condition,
      uses: this.uses,
      with: cleanObject(this.parameters),
      env: this.env,
      continueOnError: this.continueOnError,
      timeoutMinutes: this.timeoutMinutes,
    };
  }
}

/**
 * Parses the external action name into its components: owner, repo, path, and ref.
 *
 * The expected format can be either:
 * - `{owner}/{repo}@{ref}` or
 * - `{owner}/{repo}/{path}@{ref}`
 *
 * @param name - The string representation of the external action name.
 * @returns An object containing the owner, repo, path (if present), and ref.
 * @throws Error if the input string is not in a valid format.
 */
export function parseExternalActionName(name: string): {
  owner: string;
  repo: string;
  ref: string;
  path?: string;
} {
  const regex = /^([^\/]+)\/([^@\/]+)(\/(.+))?@(.+)$/;
  const match = name.match(regex);

  if (!match) {
    throw new Error(`Invalid repository reference: ${name}`);
  }

  const owner = match[1]; // The owner (e.g., "octocat")
  const repo = match[2]; // The repository name (e.g., "Hello-World")
  const path = match[4]; // The path (if it exists, otherwise undefined)
  const ref = match[5]; // The reference (e.g., "main", "v1.0", "a1b2c3d")

  // Validate the reference (you can customize this validation as needed)
  if (/[^a-zA-Z0-9._-]/.test(ref)) {
    throw new Error(`Invalid repository reference: ${name}`);
  }

  return { owner, repo, ref, path };
}

export function parseDockerActionName(name: string): { host: string; image: string; tag: string } {
  const regex = /^docker:\/\/([^\/]+)\/([^:]+)(?::(.*))?$/;
  const match = name.match(regex);

  if (!match) {
    throw new Error(`Invalid Docker action name: ${name}`);
  }

  const host = match[1]; // The host (e.g., docker.io)
  const image = match[2]; // The image name (e.g., library/ubuntu)
  const tag = match[3] || "latest"; // The tag (e.g., 20.04), defaults to 'latest' if not provided

  return { host, image, tag };
}
