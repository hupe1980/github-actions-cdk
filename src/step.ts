import type { IConstruct } from "constructs";
import { type ShellType, validShells } from "./common";
import { Component } from "./component";
import { cleanObject } from "./private/utils";

// Unique symbols to mark instances of StepBase, RunStep, and RegularStep
const STEP_BASE_SYMBOL = Symbol.for("github-actions-cdk.StepBase");
const RUN_STEP_SYMBOL = Symbol.for("github-actions-cdk.RunStep");
const REGULAR_STEP_SYMBOL = Symbol.for("github-actions-cdk.RegularStep");

/**
 * Common properties for all step types in a GitHub Actions workflow job.
 */
export interface CommonStepProps {
  /**
   * A descriptive name for the step, displayed in the GitHub Actions UI.
   * @example "Install dependencies"
   */
  readonly name?: string;

  /**
   * Conditional expression to determine if the step should run (equivalent to `if` in GitHub Actions).
   * Supports GitHub Actions expressions, e.g., `${{ success() }}`.
   * @example "${{ github.event_name == 'push' }}"
   */
  readonly condition?: string;

  /**
   * Maximum execution time for the step, in minutes.
   * @example 10
   */
  readonly timeoutMinutes?: number;

  /**
   * Whether the job should continue if this step fails.
   * @default false
   */
  readonly continueOnError?: boolean;

  /**
   * Environment variables specific to this step, overriding job-level or workflow-level variables.
   * @example { "NODE_ENV": "production" }
   */
  readonly env?: Record<string, string>;
}

/**
 * Configuration for a step that runs a shell command.
 */
export interface RunStepProps extends CommonStepProps {
  /**
   * Commands or scripts to execute in this step.
   * @example ["npm", "install"]
   */
  readonly run: string | string[];

  /**
   * Directory in which the step's command or action executes.
   * Defaults to the job's working directory if not specified.
   * @example "src/"
   */
  readonly workingDirectory?: string;

  /**
   * Shell environment for this step, allowing custom shells like `bash`, `pwsh`, etc.
   * @default "bash"
   * @example "pwsh"
   */
  readonly shell?: ShellType;
}

/**
 * Configuration for a step that uses a predefined GitHub Action.
 */
export interface RegularStepProps extends CommonStepProps {
  /**
   * GitHub Action to run, identified by repository or Docker image reference.
   * @example "actions/checkout@v2"
   */
  readonly uses: string;

  /**
   * Input parameters for the action, passed as a key-value map.
   * @example { "token": "${{ secrets.GITHUB_TOKEN }}" }
   */
  readonly parameters?: Record<string, unknown>;
}

/**
 * Base class representing a single step within a GitHub Actions job.
 */
export abstract class StepBase extends Component {
  /**
   * Checks if an object is an instance of `StepBase`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `StepBase`; otherwise, `false`.
   */
  public static isStepBase(x: unknown): x is StepBase {
    return x !== null && typeof x === "object" && STEP_BASE_SYMBOL in x;
  }

  public readonly name?: string; // Name of the step
  public readonly condition?: string; // Conditional execution expression
  public readonly env?: Record<string, string>; // Environment variables for this step
  public readonly continueOnError?: boolean; // Continue on error flag
  public readonly timeoutMinutes?: number; // Step timeout in minutes

  /**
   * Creates a new `StepBase` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this step.
   */
  constructor(scope: IConstruct, id: string, props: CommonStepProps) {
    super(scope, id);

    // Mark the construct scope with STEP_BASE_SYMBOL to denote it's a StepBase
    Object.defineProperty(this, STEP_BASE_SYMBOL, { value: true });

    this.name = props.name;
    this.condition = props.condition;
    this.env = props.env;
    this.continueOnError = props.continueOnError;
    this.timeoutMinutes = props.timeoutMinutes;
  }

  /**
   * Retrieves the step's unique identifier within the context of a workflow job.
   * @returns The unique identifier for this step.
   */
  public get id(): string {
    return this.node.id;
  }
}

/**
 * Step that runs shell commands in a GitHub Actions job.
 */
export class RunStep extends StepBase {
  /**
   * Checks if an object is an instance of `RunStep`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `RunStep`; otherwise, `false`.
   */
  public static isRunStep(x: unknown): x is RunStep {
    return x !== null && typeof x === "object" && RUN_STEP_SYMBOL in x;
  }

  public readonly run: string[]; // Commands to execute in the step
  public readonly workingDirectory?: string; // Directory for execution
  public readonly shell?: ShellType; // Shell type for execution

  /**
   * Creates a new `RunStep` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this run step.
   */
  constructor(scope: IConstruct, id: string, props: RunStepProps) {
    super(scope, id, props);

    Object.defineProperty(this, RUN_STEP_SYMBOL, { value: true });

    // Ensure the run property is always an array
    this.run = typeof props.run === "string" ? [props.run] : props.run;
    this.workingDirectory = props.workingDirectory;
    this.shell = props.shell;

    // Validation for the shell type
    this.node.addValidation({
      validate: () => {
        const errors: string[] = [];
        if (this.shell && !validShells.includes(this.shell as ShellType)) {
          errors.push(`'shell' must be one of the following: ${validShells.join(", ")}.`);
        }
        return errors;
      },
    });
  }

  /**
   * Serializes the `RunStep` configuration to an object for GitHub Actions YAML output.
   *
   * Converts the run step configuration into an object format suitable for
   * GitHub Actions YAML serialization.
   *
   * @returns An object suitable for YAML serialization.
   * @internal
   * @override
   */
  public _toRecord(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      if: this.condition,
      run: this.run.join("\n"),
      env: this.env,
      continueOnError: this.continueOnError,
      timeoutMinutes: this.timeoutMinutes,
      workingDirectory: this.workingDirectory,
      shell: this.shell,
    };
  }
}

/**
 * Step that runs a predefined GitHub Action within a job.
 */
export class RegularStep extends StepBase {
  /**
   * Checks if an object is an instance of `RegularStep`.
   *
   * @param x - The object to check.
   * @returns `true` if `x` is a `RegularStep`; otherwise, `false`.
   */
  public static isRegularStep(x: unknown): x is RegularStep {
    return x !== null && typeof x === "object" && REGULAR_STEP_SYMBOL in x;
  }

  public readonly uses: string; // Reference to the GitHub Action
  public readonly parameters?: Record<string, unknown>; // Input parameters for the action

  /**
   * Creates a new `RegularStep` instance.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The unique identifier for this step.
   * @param props - Configuration properties for this regular step.
   */
  constructor(scope: IConstruct, id: string, props: RegularStepProps) {
    super(scope, id, props);

    Object.defineProperty(this, REGULAR_STEP_SYMBOL, { value: true });

    this.uses = props.uses;
    this.parameters = props.parameters;
  }

  /**
   * Determines if the action is an external action.
   *
   * @returns `true` if the action is an external action; otherwise, `false`.
   */
  public isExternalAction(): boolean {
    return !this.isRepoAction() && !this.isDockerAction();
  }

  /**
   * Determines if the action is a repository action.
   *
   * @returns `true` if the action is a repository action; otherwise, `false`.
   */
  public isRepoAction(): boolean {
    return this.uses.startsWith("./");
  }

  /**
   * Determines if the action is a Docker action.
   *
   * @returns `true` if the action is a Docker action; otherwise, `false`.
   */
  public isDockerAction(): boolean {
    return this.uses.startsWith("docker://");
  }

  /**
   * Serializes the `RegularStep` configuration to an object for GitHub Actions YAML output.
   *
   * Converts the step configuration into an object format suitable for
   * GitHub Actions YAML serialization.
   *
   * @returns An object suitable for YAML serialization.
   * @internal
   * @override
   */
  public _toRecord(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      if: this.condition,
      uses: this.uses,
      with: cleanObject(this.parameters),
      env: this.env,
      continueOnError: this.continueOnError,
      timeoutMinutes: this.timeoutMinutes,
    };
  }
}

export function parseExternalActionName(name: string): {
  owner: string;
  repo: string;
  ref: string;
} {
  const regex = /^([^\/]+)\/([^@]+)@(.+)$/;
  const match = name.match(regex);

  if (!match) {
    throw new Error(`Invalid repository reference: ${name}`);
  }

  const owner = match[1]; // The owner (e.g., "octocat")
  const repo = match[2]; // The repository name (e.g., "Hello-World")
  const ref = match[3]; // The reference (e.g., "main", "v1.0", "a1b2c3d")

  return { owner, repo, ref };
}

export function parseDockerActionName(name: string): { host: string; image: string; tag: string } {
  const regex = /^docker:\/\/([^\/]+)\/([^:]+)(?::(.*))?$/;
  const match = name.match(regex);

  if (!match) {
    throw new Error(`Invalid Docker action name: ${name}`);
  }

  const host = match[1]; // The host (e.g., docker.io)
  const image = match[2]; // The image name (e.g., library/ubuntu)
  const tag = match[3] || "latest"; // The tag (e.g., 20.04), defaults to 'latest' if not provided

  return { host, image, tag };
}
