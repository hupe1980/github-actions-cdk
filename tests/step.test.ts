// tests/step.test.ts

import { RootConstruct } from "constructs";
import { cleanObject } from "../src/private/utils"; // Ensure this utility is properly imported
import { RegularStep, RunStep } from "../src/step"; // Adjust the import path as necessary

describe("RunStep", () => {
  let mockScope: RootConstruct;

  beforeEach(() => {
    mockScope = new RootConstruct();
  });

  it("should initialize with given properties", () => {
    const step = new RunStep(mockScope, "step1", {
      name: "Install dependencies",
      condition: "${{ success() }}",
      timeoutMinutes: 10,
      continueOnError: false,
      env: { NODE_ENV: "production" },
      run: ["npm install"],
      workingDirectory: "src/",
      shell: "bash",
    });

    expect(step.name).toBe("Install dependencies");
    expect(step.condition).toBe("${{ success() }}");
    expect(step.timeoutMinutes).toBe(10);
    expect(step.continueOnError).toBe(false);
    expect(step.env).toEqual({ NODE_ENV: "production" });
    expect(step.run).toEqual(["npm install"]);
    expect(step.workingDirectory).toBe("src/");
    expect(step.shell).toBe("bash");
  });

  it("should serialize RunStep to record format", () => {
    const step = new RunStep(mockScope, "step1", {
      name: "Run tests",
      run: ["npm test"],
      env: { NODE_ENV: "test" },
    });

    const record = step._toRecord();

    expect(record).toEqual({
      id: "step1",
      name: "Run tests",
      if: undefined,
      run: "npm test",
      env: { NODE_ENV: "test" },
      continueOnError: undefined,
      timeoutMinutes: undefined,
      workingDirectory: undefined,
      shell: undefined,
    });
  });
});

describe("RegularStep", () => {
  let mockScope: RootConstruct;

  beforeEach(() => {
    mockScope = new RootConstruct();
  });

  it("should initialize with given properties", () => {
    const step = new RegularStep(mockScope, "step2", {
      name: "Checkout code",
      uses: "actions/checkout@v2",
      parameters: { token: "${{ secrets.GITHUB_TOKEN }}" },
    });

    expect(step.name).toBe("Checkout code");
    expect(step.uses).toBe("actions/checkout@v2");
    expect(step.parameters).toEqual({ token: "${{ secrets.GITHUB_TOKEN }}" });
  });

  it("should serialize RegularStep to record format", () => {
    const step = new RegularStep(mockScope, "step2", {
      name: "Checkout code",
      uses: "actions/checkout@v2",
      parameters: { token: "${{ secrets.GITHUB_TOKEN }}" },
    });

    const record = step._toRecord();

    expect(record).toEqual({
      id: "step2",
      name: "Checkout code",
      if: undefined,
      uses: "actions/checkout@v2",
      with: cleanObject({ token: "${{ secrets.GITHUB_TOKEN }}" }),
      env: undefined,
      continueOnError: undefined,
      timeoutMinutes: undefined,
    });
  });
});
