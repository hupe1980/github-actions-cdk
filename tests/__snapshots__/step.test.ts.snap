// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step should create a step with all properties 1`] = `
{
  "continueOnError": true,
  "env": {
    "NODE_ENV": "production",
  },
  "id": "TestStepWithAllProps",
  "if": "success()",
  "name": "Build Project",
  "run": undefined,
  "shell": undefined,
  "timeoutMinutes": 10,
  "uses": "actions/setup-node@v2",
  "with": {
    "version": "14",
  },
  "workingDirectory": undefined,
}
`;

exports[`Step should create a step with minimal properties 1`] = `
{
  "continueOnError": undefined,
  "env": undefined,
  "id": "TestStep",
  "if": undefined,
  "name": "Install dependencies",
  "run": "npm install",
  "shell": undefined,
  "timeoutMinutes": undefined,
  "uses": undefined,
  "with": undefined,
  "workingDirectory": undefined,
}
`;
